### PR TITLE
Automatically add out kwarg & member & inplace_member

### PR DESCRIPTION
**Pull request by mavenlin**
As per our discussion in #18, 
- For functions that support `out=None` kwarg, we use a preset template in the decorator.
- Automatically set the member functions when they exist to avoid manually add them at two places. When we implement `torch.add`, `torch.Tensor.add` and `torch.Tensor.add_` are automatically added. While it works for most of the functions, for some of the functions this can be turned off, e.g. `torch.permute` & `torch.Tensor.permute` has different signatures.

### CODE REVIEW

File: torch2jax/__init__.py [hunk 1]
```diff
@@ -230,20 +203,46 @@ def _args_to_shape(args):
   return args if isinstance(args[0], int) else args[0]
 
 
-def implements(torch_function, Torchishify_output=True):
+def implements(torch_function, Torchishify_output=True, out_kwarg=False, Torchish_member=False):
   """Register a torch function override"""
+  if out_kwarg:
+    # adding out kwarg will by default Torchishify the output, so we don't need to Torchishify the output again.
+    Torchishify_output = False
```
[the remark this answers is not in the document]
Comment (mavenlin) [reply torch2jax/__init__.py]: what I actually mean is that `out_kwarg` already imply Torchishify output, so this is to avoid double doing it. But I see this could be a source of confusion, let me write it in another way.

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -76,6 +76,7 @@ def j2t_array(jax_array):
 
 
 HANDLED_FUNCTIONS = {}
+FUNCTIONS_WITH_OUT_KWARG = {}
```
Comment (samuela) [inline torch2jax/__init__.py:79]: why is it necessary to maintain this mapping?

File: torch2jax/__init__.py [hunk 3]
```diff
@@ -1167,7 +1164,8 @@ def f_jax(*args, rng=None):
       with RngPooperContext(None if rng is None else RngPooper(rng)):
         with TorchishMode():
           out = f(*torch_args)
-    return out.value
+    # use the torch's tree_map, because out is generated from torch code
+    return torch.utils._pytree.tree_map(lambda x: x.value if isinstance(x, Torchish) else x, out)
```
Comment (samuela) [inline torch2jax/__init__.py:1168]: why is this necessary?

Comment (mavenlin) [reply torch2jax/__init__.py:1171]: While we already have `t2j` and `j2t` that performs real conversions (between `jnp.array` to `torch.Tensor`). 

We kind of need functions to do **mock** conversion, aka, from `jnp.array/np.array` to `Torchish` and vice versa. We want to handle all types besides the arrays, i.e. `int`, `float`, `None` are kept as it is, `jnp.dtype` is converted to `torch.dtype`. And we need to handle pytree.

`_coerce` is already fulfilling the `mock_j2t`, we can add more types to it.

for the `mock_t2j`, we need to convert from `Torchish` back to `jnp.array`. Currently the implementation assumes a scalar output and simply does a unwrapping by `out.value`. But we probably need to handle pytree here, for example `max`, `topk` returns a custom type containing `(indices, values)`, which is registered with the torch's pytree util and we can unwrap with `torch.utils._pytree.tree_map`.

There are other functions that returns even more complex pytree, for example in transformers library [here](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_outputs.py#L658). We probably need to do some [registering](https://docs.jax.dev/en/latest/_autosummary/jax.tree_util.register_dataclass.html) to make it work.

But there's a problem too, because jax's tree functions do not recognize the torch's custom return data types.

**Comment by samuela** [comment]
yes, sorry about the ci. it should be fixed after b9588259ec7a2f63eed67d0c434086894a74c3e6

File: torch2jax/__init__.py (hunk 3 above)
Comment (mavenlin) [reply torch2jax/__init__.py:1171]: In the context of this PR, `torch.unbind` generates a list of tensor.
Now it can be tested as it is, without needing to wrap with lambda, which makes automatic testing of `torch.Tensor.unbind` & `torch.Tensor.unbind_` possible. I would leave it for future PRs to handle general pytrees/dataclasses.

**Comment by mavenlin** [comment]
Great, tests are passing after the ci fix. Let me know if you have further comments on this one.

**Comment by samuela** [comment]
looks like there are some merge conflicts. i'm also gonna have to meditate a bit on how exactly i want to design this. what u have is great, but i'm concerned we may eventually blow up the complexity of `implements`

**Comment by mavenlin** [comment]
> looks like there are some merge conflicts. i'm also gonna have to meditate a bit on how exactly i want to design this. what u have is great, but i'm concerned we may eventually blow up the complexity of `implements`

File: torch2jax/__init__.py (hunk 3 above)
Comment (samuela) [reply torch2jax/__init__.py:1171]: ah i see. i guess unbind wasn't getting tested before so it slipped through the cracks

File: tests/utils.py [hunk 4]
```diff
@@ -25,40 +32,105 @@ def test_t2j_array():
   aac(t2j(torch.eye(3).unsqueeze(0)), jnp.eye(3)[jnp.newaxis, ...])
 
 
+def out_kwarg_test(f, args, kwargs={}, **assert_kwargs):
+  torch_kwargs = jax.tree_util.tree_map(lambda x: j2t(x) if isinstance(x, jnp.dtype) else x, kwargs)
```
Comment (samuela) [inline tests/utils.py:36]:
```suggestion
  torch_kwargs = jax.tree.map(lambda x: j2t(x) if isinstance(x, jnp.dtype) else x, kwargs)
```
nit

File: tests/utils.py [hunk 5]
```diff
@@ -25,40 +32,105 @@ def test_t2j_array():
...
+
+  jax_function = t2j(torch_function)
+  jax_out1, jax_out2, jax_out3 = jax_function(*args)
+  assert jax_out2 is jax_out3
+  aac(jax_out1, jax_out3, **assert_kwargs)
+
+
+def Torchish_member_test(f, args, kwargs={}, inplace=False, **assert_kwargs):
+  name = f.__name__ if not inplace else f"{f.__name__}_"
+  torch_kwargs = jax.tree_util.tree_map(lambda x: j2t(x) if isinstance(x, jnp.dtype) else x, kwargs)
```
Comment (samuela) [inline tests/utils.py:52]:
```suggestion
  torch_kwargs = jax.tree.map(lambda x: j2t(x) if isinstance(x, jnp.dtype) else x, kwargs)
```

**Comment by mavenlin** [comment]
## 2025-08-19
- Explicit control what to test for each function
- Test gradient for all cases by reducing output to scalar.
- Fix gradient of `batch_norm`
